`==` in tables should not raise KeyError

flyx opened the pull request:
With previous code, the compiler deduced that equalsImpl may raise
a KeyError. While this could only actually happen in a nasty
multi-threaded environment, I fixed the code so that it will never happen.

Example code:

```nim
import tables

proc tablesAreEqual(left, right: Table[string, string]): bool {.raises: [].} =
  left == right

let
  a = {"a": "b"}.newTable()
  b = {"c": "d"}.newTable()

echo tablesAreEqual(a, b)
```

Did not work previously, but works now.

Araq commented:
This solution is too expensive. Better use ``getOrDefault`` instead which cannot raise (right?).

flyx commented:
Modified it. Curiously, now automatic dereference did not work anymore. I got:

```
CC: stdlib_tables
Error: execution of an external compiler program 'clang -c  -w  -I/Users/flyx/Projects/3rdParty/Nim/lib -o /Users/flyx/Projects/3rdParty/Nim/nimcache/stdlib_tables.o /Users/flyx/Projects/3rdParty/Nim/nimcache/stdlib_tables.c' failed with exit code: 256

/Users/flyx/Projects/3rdParty/Nim/nimcache/stdlib_tables.c:643:47: error: passing 'Table143011 *' (aka 'struct Table143011 *') to parameter of incompatible type 'Table143011' (aka 'struct Table143011'); dereference with *
                                                                        LOC29 = getordefault_143144_30748785(t0, key_143045_30748785);
                                                                                                             ^~
                                                                                                             *
/Users/flyx/Projects/3rdParty/Nim/nimcache/stdlib_tables.c:550:69: note: passing argument to parameter 't0' here
N_NIMCALL(NimStringDesc*, getordefault_143144_30748785)(Table143011 t0, NimStringDesc* key0) {
```

So I modified `equalsImpl` to take `t` as explicit parameter.

Araq commented:
Report this codegen bug properly please.

kierdavis commented:
The gcc/clang error goes away if the definition of `getOrDefault` for the `TableRef` type is moved to before the definition of `equalsImpl`. The codegen bug still needs to be fixed however.
